From d5730acf71342ba61fbd1d619ee23f451d3ce2eb Mon Sep 17 00:00:00 2001
From: Peter Chen <peter.chen@nxp.com>
Date: Wed, 10 Oct 2018 09:44:54 +0800
Subject: [PATCH 4797/5242] MLK-19601-1 xhci: Fix leaking USB3 shared_hcd at
 xhci removal

commit  0404e11ef566b3aa1d87edf8af221833eb700eea from
https://source.codeaurora.org/external/imx/linux-imx.git

(Merged upstream reviewing patch, and add cdns support -- Peter Chen)

Ensure that the shared_hcd pointer is valid when calling usb_put_hcd()

The shared_hcd is removed and freed in xhci by first calling
usb_remove_hcd(xhci->shared_hcd), and later
usb_put_hcd(xhci->shared_hcd)

Afer commit fe190ed0d602 ("xhci: Do not halt the host until both HCD have
disconnected their devices.") the shared_hcd was never properly put as
xhci->shared_hcd was set to NULL before usb_put_hcd(xhci->shared_hcd) was
called.

shared_hcd (USB3) is removed before primary hcd (USB2).
While removing the primary hcd we might need to handle xhci interrupts
to cleanly remove last USB2 devices, therefore we need to set
xhci->shared_hcd to NULL before removing the primary hcd to let xhci
interrupt handler know shared_hcd is no longer available.

xhci-plat.c, cdns/host.c first create both their hcd's before
adding them. so to keep the correct reverse removal order use a temporary
shared_hcd variable for them.
For more details see commit 4ac53087d6d4 ("usb: xhci: plat: Create both
HCDs before adding them")

Fixes: fe190ed0d602 ("xhci: Do not halt the host until both
	       	HCD have disconnected their devices.")
Cc: Joel Stanley <joel@jms.id.au>
Cc: Chunfeng Yun <chunfeng.yun@mediatek.com>
Cc: Thierry Reding <treding@nvidia.com>
Cc: Jianguo Sun <sunjianguo1@huawei.com>
Cc: <stable@vger.kernel.org>
Tested-by: Peter Chen <peter.chen@nxp.com>
Tested-by: Jack Pham <jackp@codeaurora.org>
Reported-by: Jack Pham <jackp@codeaurora.org>
Signed-off-by: Mathias Nyman <mathias.nyman@linux.intel.com>
Signed-off-by: Peter Chen <peter.chen@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/cdns3/host.c |    9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/drivers/usb/cdns3/host.c b/drivers/usb/cdns3/host.c
index 0de3e52..37bbae1 100644
--- a/drivers/usb/cdns3/host.c
+++ b/drivers/usb/cdns3/host.c
@@ -210,16 +210,19 @@ static int cdns3_host_start(struct cdns3 *cdns)
 static void cdns3_host_stop(struct cdns3 *cdns)
 {
 	struct device *dev = cdns->host_dev;
-	struct usb_hcd	*hcd;
+	struct usb_hcd	*hcd, *shared_hcd;
 	struct xhci_hcd	*xhci;
 
 	if (dev) {
 		hcd = dev_get_drvdata(dev);
 		xhci = hcd_to_xhci(hcd);
-		usb_remove_hcd(xhci->shared_hcd);
+		shared_hcd = xhci->shared_hcd;
+		xhci->xhc_state |= XHCI_STATE_REMOVING;
+		usb_remove_hcd(shared_hcd);
+		xhci->shared_hcd = NULL;
 		usb_remove_hcd(hcd);
 		synchronize_irq(cdns->irq);
-		usb_put_hcd(xhci->shared_hcd);
+		usb_put_hcd(shared_hcd);
 		usb_put_hcd(hcd);
 		cdns->host_dev = NULL;
 		pm_runtime_set_suspended(dev);
-- 
1.7.9.5

